src/Minisite.php

```diff
@@ -1,31 +1,214 @@
 <?php
 
-/**
- * @file
- * Minisite class.
- */
-
 namespace Drupal\minisite;
 
-use Drupal\file\Entity\File;
-use Drupal\Core\Entity\EntityInterface;
-use Drupal\minisite\Minisite\MinisiteAbstract;
+use Drupal\Core\File\FileSystemInterface;
+use Drupal\file\FileInterface;
+use Drupal\minisite\Exception\AssetException;
+use Drupal\minisite\Exception\InvalidContentArchiveException;
+use Drupal\minisite\Exception\InvalidFormatArchiveException;
 
 /**
- * Class Minisite
+ * Class Minisite.
+ *
+ * Handles all interactions with Minisite data, including asset management.
+ *
  * @package Drupal\minisite
  */
-class Minisite extends MinisiteAbstract {
+class Minisite implements MinisiteInterface {
+
+  /**
+   * The archive file for this minisite.
+   *
+   * @var \Drupal\file\FileInterface
+   */
+  protected $archiveFile;
+
+  /**
+   * Array of assets for this minisite.
+   *
+   * @var \Drupal\minisite\MinisiteAsset[]
+   */
+  protected $assets = [];
+
+  /**
+   * Minisite constructor.
+   *
+   * @param \Drupal\file\FileInterface $archiveFile
+   *   The archive file.
+   * @param string $extensions
+   *   String list of archive content extensions.
+   */
+  public function __construct(FileInterface $archiveFile, $extensions) {
+    // Although archive has already been validated during upload, we still
+    // need to ensure that provided file is valid to ensure data integrity.
+    static::validateArchive($archiveFile, $extensions);
+    $this->archiveFile = $archiveFile;
+    $this->processArchive();
+  }
+
+  /**
+   * Instantiate this class from the archive file.
+   *
+   * @param \Drupal\file\FileInterface $archiveFile
+   *   The file to instantiate the class.
+   * @param string $extensions
+   *   String list of archive content extensions.
+   *
+   * @return \Drupal\minisite\Minisite
+   *   An instance of this class.
+   */
+  public static function fromArchive(FileInterface $archiveFile, $extensions) {
+    $instance = new self($archiveFile, $extensions);
+
+    return $instance;
+  }
+
   /**
-   * Presave minisite assets.
+   * Set archive file.
    *
-   * @param \Drupal\file\Entity\File $file
-   * @param \Drupal\Core\Entity\EntityInterface $entity
-   * @return bool|string
+   * @param \Drupal\file\FileInterface $file
+   *   Already uploaded archive file object to set.
    */
-  static public function preSave(File $file, EntityInterface $entity) {
-    $minisite_asset_path = self::extractAsset($file);
+  public function setArchiveFile(FileInterface $file) {
+    $this->archiveFile = $file;
+  }
 
-    return !empty($minisite_asset_path) ? $minisite_asset_path : FALSE;
+  /**
+   * Get archive file.
+   *
+   * @return \Drupal\file\FileInterface
+   *   Archive file used to instantiate this minisite.
+   */
+  public function getArchiveFile() {
+    return $this->archiveFile;
   }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function validateArchive(FileInterface $file, $contentExtensions) {
```

Comment:
note that this function can be called with file being uploaded (file that has a temporary URI) and file that has already been uploaded (file leaving in the site's `files`).